sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+emit some additional semantic checking. Shall we leave it in place or that
+request for semantic checking should also be documented by
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute when `kind`
+is set to a certain value?
+
+**TODO**: alternatively, we could completely re-use existing
+`[[__sycl_detail__::device_global]]` attribute and introduce another one for
+specialization constants, i.e. it is a question of whether or not we want to
+generalize unique IDs generation in form of a generic attribute or not.
+
+When DPC++ compiler is used as both host and device compiler, then the attribute
+should be respected by both host and device compiler passes and LLVM IR
+attributes should appear in LLVM IR for both host and device code. When DPC++
+compiler is only used as a device compiler, then we don't expect the attribute
+to be handled on host, apparently.
+
+Another thing we need from DPC++ FE compiler is to define a special macro, which
+will allow to distinguish it from other compilers. That is needed to apply the
+aforementioned attribute conditionally to avoid spamming users with warnings
+about unknown attributes.
+
+The suggested macro name is `__INTEL_SYCL_HOST_COMPILER__`. It should be defined
+when the compiler is invoked in SYCL host mode (`-fsycl-is-host` `-cc1` flag).
+
+## Common headers part
+
+Header files should be modified by adding the new attributes to types
+declarations, objects of which we will need in our mapping.Again,
+`device_global` and `specialization_id` are examples here:
+
+```
+template <typename T>
+class
+#if defined(__SYCL_DEVICE_ONLY__) || defined(__INTEL_SYCL_HOST_COMPILER__)
+  [[__sycl_detail__::uniquely_identifiable_object("specialization_id")]]
+#endif
+specialization_id {
+// ...
+};
+```
+
+## Common runtime part
+
+The runtime should implement the following function, which will be called from
+a code generated by the compiler (see the next section):
+
+```
+void __register_uniquely_identifiable_object(
+  void *Address, const char* UniqueID, const char *Kind);
+```
+
+The function accepts the following arguments:
+- `Address` is an address of a variable, which exists in an application on host
+- `UniqueID` is a unique symbolic ID, which corresponds to that variable
+- `Kind` is a string which corresponds to `kind` argument passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute attached
+  to the type of the variable identified by `Address`. It can be used to
+  distinguish different entities like `specialization_id` and `device_global`:
+  for example they could be stored in different maps to speed up certain
+  operations with them.
+
+The compiler guarantees that the function will be called zero or more times
+(depending on the amount of uniquely identifiable objects found in a program)
+_before_ application's `main()` function, i.e. in a global constructor.
+
+That poses some restrictions on those uniquely identifiable object, i.e. that
```

Comment:
```suggestion
That poses some restrictions on those uniquely identifiable objects, i.e. that
```

Reply:
Applied in 6c7a9be71e57c2855e87e18de145a833629dc060